File: asap-planner/utils/logics.py
import copy
import math
from loguru import logger

from promql_utilities.data_model.KeyByLabelNames import KeyByLabelNames
from promql_utilities.query_logics.enums import QueryPatternType, CleanupPolicy
from promql_utilities.ast_matching.PromQLPattern import MatchResult
from promql_utilities.query_logics.logics import (
    does_precompute_operator_support_subpopulations,
)

CMS_WITH_HEAP_MULT = 4

# Default sketch parameters for backward compatibility
DEFAULT_SKETCH_PARAMETERS = {
    "CountMinSketch": {"depth": 3, "width": 1024},
    "CountMinSketchWithHeap": {"depth": 3, "width": 1024, "heap_multiplier": 4},
    "DatasketchesKLL": {"K": 20},
    "HydraKLL": {"row_num": 3, "col_num": 1024, "k": 20},
}


def get_effective_repeat(t_repeat: int, step: int) -> int:
    """
    Calculate effective repeat interval for range queries.

    For range queries (step > 0), use the smaller of t_repeat and step to ensure
    we produce aggregates frequently enough to support the query step size.
    For instant queries (step = 0), use t_repeat.
    """
    return min(t_repeat, step) if step > 0 else t_repeat


# TODO:
# We only show the logic of `get_precompute_operator_parameters` here.
# Semantics for topk query will be added in later PRs.
def get_precompute_operator_parameters(
    aggregation_type: str,
    aggregation_sub_type: str,
    query_pattern_match: MatchResult,
    sketch_parameters: dict,
) -> dict:
    # Allow partial overrides: use provided parameters, fall back to defaults per sketch type
    if sketch_parameters is None:
        sketch_parameters = {}

    if aggregation_type in [
        "Increase",
        "MinMax",
        "Sum",
        "MultipleIncrease",
        "MultipleMinMax",
        "MultipleSum",
        "DeltaSetAggregator",
        "SetAggregator",
    ]:
        return {}
    elif aggregation_type == "CountMinSketch":
        params = sketch_parameters.get(
            "CountMinSketch", DEFAULT_SKETCH_PARAMETERS["CountMinSketch"]
        )
        return {"depth": params["depth"], "width": params["width"]}
    elif aggregation_type == "CountMinSketchWithHeap":
        if aggregation_sub_type == "topk":
            if "aggregation" not in query_pattern_match.tokens:
                raise ValueError(
                    f"{aggregation_sub_type} query missing aggregator in the match tokens"
                )
            if "param" not in query_pattern_match.tokens["aggregation"]:
                raise ValueError(
                    f"{aggregation_sub_type} query missing required 'k' parameter"
                )
            k = int(query_pattern_match.tokens["aggregation"]["param"].val)
            params = sketch_parameters.get(
                "CountMinSketchWithHeap",
                DEFAULT_SKETCH_PARAMETERS["CountMinSketchWithHeap"],
            )
            heap_mult = params.get("heap_multiplier", CMS_WITH_HEAP_MULT)
            return {
                "depth": params["depth"],
                "width": params["width"],
                "heapsize": k * heap_mult,
            }
        else:
            raise ValueError(
                f"Aggregation sub-type {aggregation_sub_type} for CountMinSketchWithHeap not supported"
            )
    elif aggregation_type == "DatasketchesKLL":
        params = sketch_parameters.get(
            "DatasketchesKLL", DEFAULT_SKETCH_PARAMETERS["DatasketchesKLL"]
        )
        return {"K": params["K"]}
    elif aggregation_type == "HydraKLL":
        params = sketch_parameters.get(
            "HydraKLL", DEFAULT_SKETCH_PARAMETERS["HydraKLL"]
        )
        return {
            "row_num": params["row_num"],
            "col_num": params["col_num"],
            "k": params["k"],
        }
    # elif aggregation_type == "UnivMon":
    #     return {"depth": 3, "width": 2048, "levels": 16}
    else:
        raise NotImplementedError(f"Aggregation type {aggregation_type} not supported")


def get_cleanup_param(
    cleanup_policy: CleanupPolicy,
    query_pattern_type,
    query_pattern_match,
    t_repeat: int,
    window_type: str,
    range_duration: int,
    step: int,
) -> int:
    """
    Calculate cleanup parameter based on cleanup policy and range query params.

    Sliding windows (both policies): range_duration / step + 1
    Tumbling circular_buffer: (T_lookback + range_duration) / min(T_repeat, step)
    Tumbling read_based: (T_lookback / min(T_repeat, step)) * (range_duration / step + 1)

    For ONLY_SPATIAL queries, T_lookback = T_repeat.
    For instant queries, range_duration = 0 and effective_repeat = T_repeat.

    Args:
        cleanup_policy: CleanupPolicy.CIRCULAR_BUFFER or CleanupPolicy.READ_BASED
        query_pattern_type: QueryPatternType enum
        query_pattern_match: MatchResult with query tokens
        t_repeat: Query repeat interval in seconds
        window_type: "sliding" or "tumbling"
        range_duration: end - start in seconds (0 for instant queries)
        step: Range query step in seconds (required if range_duration > 0)

    Raises:
        ValueError: If exactly one of range_duration or step is zero
    """
    # Validation: range_duration and step must both be zero (instant) or both non-zero (range)
    if (range_duration == 0) != (step == 0):
        raise ValueError(
            f"range_duration and step must both be 0 (instant query) or both > 0 (range query). "
            f"Got range_duration={range_duration}, step={step}"
        )

    is_range_query = step > 0

    # For ONLY_SPATIAL, T_lookback = T_repeat
    if query_pattern_type == QueryPatternType.ONLY_SPATIAL:
        t_lookback = t_repeat
    else:
        t_lookback = int(
            query_pattern_match.tokens["range_vector"]["range"].total_seconds()
        )

    # For sliding windows: range_duration / step + 1 (same for both policies)
    if window_type == "sliding":
        if is_range_query:
            result = range_duration // step + 1
        else:
            result = 1  # instant query
        logger.debug(
            f"Sliding window mode: cleanup_param = {result} "
            f"(range_duration={range_duration}s, step={step}s)"
        )
        return result

    # Tumbling window calculations
    effective_repeat = get_effective_repeat(t_repeat, step)

    # We use ceiling division because even if the time span doesn't fully fill
    # a bucket, we still need that bucket to cover the partial data.
    # E.g., if T_lookback=10s and effective_repeat=100s, we still need 1 bucket.
    if cleanup_policy == CleanupPolicy.CIRCULAR_BUFFER:
        # ceil((T_lookback + range_duration) / effective_repeat)
        result = math.ceil((t_lookback + range_duration) / effective_repeat)
    elif cleanup_policy == CleanupPolicy.READ_BASED:
        # ceil(T_lookback / effective_repeat) * (range_duration / step + 1)
        lookback_buckets = math.ceil(t_lookback / effective_repeat)
        if is_range_query:
            num_steps = range_duration // step + 1
        else:
            num_steps = 1  # instant query
        result = lookback_buckets * num_steps
    else:
        raise ValueError(f"Invalid cleanup policy: {cleanup_policy}")

    logger.debug(
        f"Tumbling window mode ({cleanup_policy.value}): cleanup_param = {result} "
        f"(t_lookback={t_lookback}s, t_repeat={t_repeat}s, "
        f"range_duration={range_duration}s, step={step}s)"
    )
    return result


def should_use_sliding_window(query_pattern_type, aggregation_type):
    """
    Decide if sliding windows should be used based on query type and aggregation type.

    For Issue #236: Use sliding windows for ALL ONLY_TEMPORAL queries except DeltaSetAggregator.
    This eliminates merging overhead in QueryEngine at the cost of more computation in Arroyo.

    Args:
        query_pattern_type: ONLY_TEMPORAL, ONLY_SPATIAL, or ONE_TEMPORAL_ONE_SPATIAL
        aggregation_type: Type of aggregation (e.g., 'DatasketchesKLL', 'Sum', etc.)

    Returns:
        bool: True if sliding windows should be used
    """
    # NOTE: returning False since sliding window pipelines are causing arroyo to crash
    return False
    # Only use sliding for ONLY_TEMPORAL queries (not ONE_TEMPORAL_ONE_SPATIAL or ONLY_SPATIAL)
    if query_pattern_type != QueryPatternType.ONLY_TEMPORAL:
        logger.debug(
            f"Query pattern {query_pattern_type} not eligible for sliding windows "
            f"(only ONLY_TEMPORAL supported)"
        )
        return False

    # Explicitly exclude DeltaSetAggregator (paired with CMS but needs tumbling)
    if aggregation_type == "DeltaSetAggregator":
        logger.debug("DeltaSetAggregator excluded from sliding windows")
        return False

    # All other ONLY_TEMPORAL aggregations use sliding windows
    logger.info(
        f"Aggregation type '{aggregation_type}' with {query_pattern_type} -> SLIDING windows"
    )
    return True


def set_window_parameters(
    query_pattern_type,
    query_pattern_match,
    t_repeat,
    prometheus_scrape_interval,
    aggregation_type,
    template_config,
    step: int,
):
    """
    Set window parameters for streaming aggregation config.
    Auto-decides between sliding and tumbling windows based on query type and aggregation cost.

    For ONLY_TEMPORAL queries with expensive aggregations (KLL, CMS):
    - Uses SLIDING windows: windowSize = range duration, slideInterval = effective_repeat
    - This reduces QueryEngine latency by avoiding merges (Arroyo does more work upfront)

    For other queries:
    - Uses TUMBLING windows: windowSize = slideInterval = effective_repeat
    - This is the original behavior

    For range queries (step > 0), effective_repeat = min(t_repeat, step).
    For instant queries (step = 0), effective_repeat = t_repeat.

    Args:
        query_pattern_type: Pattern type (ONLY_TEMPORAL, ONLY_SPATIAL, ONE_TEMPORAL_ONE_SPATIAL)
        query_pattern_match: Matched PromQL pattern containing query metadata
        t_repeat: Query repeat interval in seconds
        prometheus_scrape_interval: Scrape interval in seconds
        aggregation_type: Type of aggregation operator
        template_config: StreamingAggregationConfig to update
        step: Range query step in seconds (0 for instant queries)
    """
    # For range queries, use min(t_repeat, step) as the effective repeat interval
    effective_repeat = get_effective_repeat(t_repeat, step)

    # Decide if we should use sliding windows
    use_sliding_window = should_use_sliding_window(query_pattern_type, aggregation_type)

    if use_sliding_window:
        # SLIDING WINDOW for ONLY_TEMPORAL queries with expensive aggregations
        logger.info(
            f"Configuring SLIDING WINDOW for {query_pattern_type} "
            f"with {aggregation_type}"
        )

        if query_pattern_type == QueryPatternType.ONLY_TEMPORAL:
            # Window size = range duration (e.g., 15m = 900s)
            range_seconds = int(
                query_pattern_match.tokens["range_vector"]["range"].total_seconds()
            )

            # Check if this is actually a tumbling window (windowSize == slideInterval)
            if range_seconds == effective_repeat:
                logger.info(
                    f"Detected windowSize == slideInterval ({range_seconds}s). "
                    f"Using tumbling window instead of sliding for efficiency."
                )
                template_config.windowSize = effective_repeat
                template_config.slideInterval = effective_repeat
                template_config.windowType = "tumbling"
                template_config.tumblingWindowSize = effective_repeat
            else:
                # True sliding window
                template_config.windowSize = range_seconds
                template_config.slideInterval = effective_repeat
                template_config.windowType = "sliding"

                logger.info(
                    f"Sliding window params: windowSize={range_seconds}s, "
                    f"slideInterval={effective_repeat}s "
                    f"(each window has {range_seconds} seconds of data, slides every {effective_repeat}s)"
                )

                # Set deprecated field for backward compatibility
                template_config.tumblingWindowSize = effective_repeat
        else:
            # This should never be reached due to should_use_sliding_window() check
            assert False, (
                f"should_use_sliding_window returned True for {query_pattern_type}, "
                f"but sliding windows only supported for ONLY_TEMPORAL"
            )
    else:
        # TUMBLING WINDOW (existing logic)
        logger.info(
            f"Configuring TUMBLING WINDOW for {query_pattern_type} "
            f"with {aggregation_type}"
        )
        _set_tumbling_window_parameters(
            query_pattern_type,
            effective_repeat,
            prometheus_scrape_interval,
            template_config,
        )


def _set_tumbling_window_parameters(
    query_pattern_type, effective_repeat, prometheus_scrape_interval, template_config
):
    """
    Set tumbling window parameters.

    Args:
        query_pattern_type: Pattern type (ONLY_TEMPORAL, ONLY_SPATIAL, ONE_TEMPORAL_ONE_SPATIAL)
        effective_repeat: Effective repeat interval (min(t_repeat, step) for range queries)
        prometheus_scrape_interval: Scrape interval in seconds
        template_config: StreamingAggregationConfig to update
    """
    if (
        query_pattern_type == QueryPatternType.ONLY_TEMPORAL
        or query_pattern_type == QueryPatternType.ONE_TEMPORAL_ONE_SPATIAL
    ):
        template_config.windowSize = effective_repeat
        template_config.slideInterval = effective_repeat
        template_config.windowType = "tumbling"
        template_config.tumblingWindowSize = effective_repeat

        logger.debug(
            f"Tumbling window params: windowSize={effective_repeat}s, slideInterval={effective_repeat}s"
        )
    elif query_pattern_type == QueryPatternType.ONLY_SPATIAL:
        template_config.windowSize = prometheus_scrape_interval
        template_config.slideInterval = prometheus_scrape_interval
        template_config.windowType = "tumbling"
        template_config.tumblingWindowSize = prometheus_scrape_interval

        logger.debug(
            f"Tumbling window params: windowSize={prometheus_scrape_interval}s, "
            f"slideInterval={prometheus_scrape_interval}s"
        )
    else:
        raise ValueError("Invalid query pattern type")


# COMMENTED OUT - Original function kept for rollback
# Issue #236: Replaced with set_window_parameters() to support sliding windows
#
# def set_tumbling_window_size(
#     query_pattern_type, t_repeat, prometheus_scrape_interval, template_config
# ):
#     if (
#         query_pattern_type == QueryPatternType.ONLY_TEMPORAL
#         or query_pattern_type == QueryPatternType.ONE_TEMPORAL_ONE_SPATIAL
#     ):
#         template_config.tumblingWindowSize = t_repeat
#     elif query_pattern_type == QueryPatternType.ONLY_SPATIAL:
#         template_config.tumblingWindowSize = prometheus_scrape_interval
#     else:
#         raise ValueError("Invalid query pattern type")


def set_subpopulation_labels(
    statistic_to_compute,
    aggregation_type,
    subpopulation_labels: KeyByLabelNames,
    template_config,
):
    if does_precompute_operator_support_subpopulations(
        statistic_to_compute, aggregation_type
    ):
        template_config.labels["grouping"] = KeyByLabelNames([])
        template_config.labels["aggregated"] = copy.deepcopy(subpopulation_labels)
    else:
        template_config.labels["grouping"] = copy.deepcopy(subpopulation_labels)
        template_config.labels["aggregated"] = KeyByLabelNames([])

File: asap-planner-rs/src/planner/logics.rs
use crate::config::input::SketchParameterOverrides;
use asap_types::enums::{CleanupPolicy, WindowType};
use promql_utilities::ast_matching::PromQLMatchResult;
use promql_utilities::data_model::KeyByLabelNames;
use promql_utilities::query_logics::enums::{AggregationType, QueryPatternType, Statistic};
use promql_utilities::query_logics::logics::does_precompute_operator_support_subpopulations;
use std::collections::HashMap;

// Default sketch parameters
const DEFAULT_CMS_DEPTH: u64 = 3;
const DEFAULT_CMS_WIDTH: u64 = 1024;
const DEFAULT_CMS_HEAP_MULT: u64 = 4;
const DEFAULT_KLL_K: u64 = 20;
const DEFAULT_HYDRA_ROW: u64 = 3;
const DEFAULT_HYDRA_COL: u64 = 1024;
const DEFAULT_HYDRA_K: u64 = 20;

pub fn get_effective_repeat(t_repeat: u64, step: u64) -> u64 {
    if step > 0 {
        t_repeat.min(step)
    } else {
        t_repeat
    }
}

pub fn should_use_sliding_window(
    _query_pattern_type: QueryPatternType,
    _aggregation_type: &str,
) -> bool {
    // HARDCODED: sliding windows crash Arroyo
    false
}

pub fn set_window_parameters(
    query_pattern_type: QueryPatternType,
    t_repeat: u64,
    prometheus_scrape_interval: u64,
    aggregation_type: &str,
    step: u64,
    config: &mut IntermediateWindowConfig,
) {
    let effective_repeat = get_effective_repeat(t_repeat, step);
    let _use_sliding = should_use_sliding_window(query_pattern_type, aggregation_type);
    // use_sliding is always false, so always tumbling
    set_tumbling_window_parameters(
        query_pattern_type,
        effective_repeat,
        prometheus_scrape_interval,
        config,
    );
}

fn set_tumbling_window_parameters(
    query_pattern_type: QueryPatternType,
    effective_repeat: u64,
    prometheus_scrape_interval: u64,
    config: &mut IntermediateWindowConfig,
) {
    match query_pattern_type {
        QueryPatternType::OnlyTemporal | QueryPatternType::OneTemporalOneSpatial => {
            config.window_size = effective_repeat;
            config.slide_interval = effective_repeat;
            config.window_type = WindowType::Tumbling;
        }
        QueryPatternType::OnlySpatial => {
            config.window_size = prometheus_scrape_interval;
            config.slide_interval = prometheus_scrape_interval;
            config.window_type = WindowType::Tumbling;
        }
    }
}

/// A mutable window config holder used during planning
#[derive(Debug, Clone, Default)]
pub struct IntermediateWindowConfig {
    pub window_size: u64,
    pub slide_interval: u64,
    pub window_type: WindowType,
}

/// Shared sketch parameter builder used by both PromQL and SQL paths.
///
/// `topk_k` is only required for `CountMinSketchWithHeap`: PromQL supplies it
/// from the `topk(k, …)` query argument; SQL passes `None` (SQL never produces
/// this operator today, so the `None` branch is unreachable in practice).
pub fn build_sketch_parameters(
    aggregation_type: AggregationType,
    aggregation_sub_type: &str,
    topk_k: Option<u64>,
    sketch_params: Option<&SketchParameterOverrides>,
) -> Result<HashMap<String, serde_json::Value>, String> {
    match aggregation_type {
        AggregationType::Increase
        | AggregationType::MinMax
        | AggregationType::Sum
        | AggregationType::MultipleIncrease
        | AggregationType::MultipleMinMax
        | AggregationType::MultipleSum
        | AggregationType::DeltaSetAggregator
        | AggregationType::SetAggregator => Ok(HashMap::new()),

        AggregationType::CountMinSketch => {
            let depth = sketch_params
                .and_then(|p| p.count_min_sketch.as_ref())
                .map(|p| p.depth)
                .unwrap_or(DEFAULT_CMS_DEPTH);
            let width = sketch_params
                .and_then(|p| p.count_min_sketch.as_ref())
                .map(|p| p.width)
                .unwrap_or(DEFAULT_CMS_WIDTH);
            let mut m = HashMap::new();
            m.insert("depth".to_string(), serde_json::Value::Number(depth.into()));
            m.insert("width".to_string(), serde_json::Value::Number(width.into()));
            Ok(m)
        }

        AggregationType::CountMinSketchWithHeap => {
            if aggregation_sub_type != "topk" {
                return Err(format!(
                    "Aggregation sub-type {} for CountMinSketchWithHeap not supported",
                    aggregation_sub_type
                ));
            }
            let k = topk_k
                .ok_or_else(|| "CountMinSketchWithHeap requires a topk k value".to_string())?;
            let depth = sketch_params
                .and_then(|p| p.count_min_sketch_with_heap.as_ref())
                .map(|p| p.depth)
                .unwrap_or(DEFAULT_CMS_DEPTH);
            let width = sketch_params
                .and_then(|p| p.count_min_sketch_with_heap.as_ref())
                .map(|p| p.width)
                .unwrap_or(DEFAULT_CMS_WIDTH);
            let heap_mult = sketch_params
                .and_then(|p| p.count_min_sketch_with_heap.as_ref())
                .and_then(|p| p.heap_multiplier)
                .unwrap_or(DEFAULT_CMS_HEAP_MULT);
            let mut m = HashMap::new();
            m.insert("depth".to_string(), serde_json::Value::Number(depth.into()));
            m.insert("width".to_string(), serde_json::Value::Number(width.into()));
            m.insert(
                "heapsize".to_string(),
                serde_json::Value::Number((k * heap_mult).into()),
            );
            Ok(m)
        }

        AggregationType::DatasketchesKLL => {
            let k = sketch_params
                .and_then(|p| p.datasketches_kll.as_ref())
                .map(|p| p.k)
                .unwrap_or(DEFAULT_KLL_K);
            let mut m = HashMap::new();
            m.insert("K".to_string(), serde_json::Value::Number(k.into()));
            Ok(m)
        }

        AggregationType::HydraKLL => {
            let row_num = sketch_params
                .and_then(|p| p.hydra_kll.as_ref())
                .map(|p| p.row_num)
                .unwrap_or(DEFAULT_HYDRA_ROW);
            let col_num = sketch_params
                .and_then(|p| p.hydra_kll.as_ref())
                .map(|p| p.col_num)
                .unwrap_or(DEFAULT_HYDRA_COL);
            let k = sketch_params
                .and_then(|p| p.hydra_kll.as_ref())
                .map(|p| p.k)
                .unwrap_or(DEFAULT_HYDRA_K);
            let mut m = HashMap::new();
            m.insert(
                "row_num".to_string(),
                serde_json::Value::Number(row_num.into()),
            );
            m.insert(
                "col_num".to_string(),
                serde_json::Value::Number(col_num.into()),
            );
            m.insert("k".to_string(), serde_json::Value::Number(k.into()));
            Ok(m)
        }

        other => Err(format!("Aggregation type {} not supported", other)),
    }
}

/// PromQL wrapper: extracts the topk `k` from the match result when needed,
/// then delegates to `build_sketch_parameters`.
pub fn build_sketch_parameters_from_promql(
    aggregation_type: AggregationType,
    aggregation_sub_type: &str,
    match_result: &PromQLMatchResult,
    sketch_params: Option<&SketchParameterOverrides>,
) -> Result<HashMap<String, serde_json::Value>, String> {
    let topk_k = if aggregation_type == AggregationType::CountMinSketchWithHeap {
        let k: u64 = match_result
            .tokens
            .get("aggregation")
            .and_then(|t| t.aggregation.as_ref())
            .and_then(|a| a.param.as_ref())
            .and_then(|p| p.parse::<f64>().ok())
            .map(|f| f as u64)
            .ok_or_else(|| "topk query missing required 'k' parameter".to_string())?;
        Some(k)
    } else {
        None
    };
    build_sketch_parameters(
        aggregation_type,
        aggregation_sub_type,
        topk_k,
        sketch_params,
    )
}

pub fn get_cleanup_param(
    cleanup_policy: CleanupPolicy,
    query_pattern_type: QueryPatternType,
    match_result: &PromQLMatchResult,
    t_repeat: u64,
    window_type: WindowType,
    range_duration: u64,
    step: u64,
) -> Result<u64, String> {
    // Validation
    if (range_duration == 0) != (step == 0) {
        return Err(format!(
            "range_duration and step must both be 0 or both > 0. Got range_duration={}, step={}",
            range_duration, step
        ));
    }

    let is_range_query = step > 0;

    let t_lookback: u64 = if query_pattern_type == QueryPatternType::OnlySpatial {
        t_repeat
    } else {
        match_result
            .get_range_duration()
            .map(|d| d.num_seconds() as u64)
            .ok_or_else(|| "No range_vector token found".to_string())?
    };

    if window_type == WindowType::Sliding {
        let result = if is_range_query {
            range_duration / step + 1
        } else {
            1
        };
        return Ok(result);
    }

    // Tumbling
    let effective_repeat = get_effective_repeat(t_repeat, step);

    let result = match cleanup_policy {
        CleanupPolicy::CircularBuffer => {
            // ceil((t_lookback + range_duration) / effective_repeat)
            let numerator = t_lookback + range_duration;
            numerator.div_ceil(effective_repeat)
        }
        CleanupPolicy::ReadBased => {
            // ceil(t_lookback / effective_repeat) * (range_duration / step + 1)
            let lookback_buckets = t_lookback.div_ceil(effective_repeat);
            let num_steps = if is_range_query {
                range_duration / step + 1
            } else {
                1
            };
            lookback_buckets * num_steps
        }
        CleanupPolicy::NoCleanup => {
            return Err("NoCleanup policy should not call get_cleanup_param".to_string());
        }
    };

    Ok(result)
}

pub fn set_subpopulation_labels(
    statistic: Statistic,
    aggregation_type: AggregationType,
    subpopulation_labels: &KeyByLabelNames,
    rollup_labels: &mut KeyByLabelNames,
    grouping_labels: &mut KeyByLabelNames,
    aggregated_labels: &mut KeyByLabelNames,
) {
    // rollup is set by caller before calling this function
    let _ = rollup_labels; // not modified here
    if does_precompute_operator_support_subpopulations(statistic, aggregation_type) {
        *grouping_labels = KeyByLabelNames::empty();
        *aggregated_labels = subpopulation_labels.clone();
    } else {
        *grouping_labels = subpopulation_labels.clone();
        *aggregated_labels = KeyByLabelNames::empty();
    }
}

/// SQL cleanup param — SQL queries are always instant (no range_duration/step).
pub fn get_sql_cleanup_param(
    cleanup_policy: CleanupPolicy,
    t_lookback: u64,
    t_repeat: u64,
) -> Result<u64, String> {
    match cleanup_policy {
        CleanupPolicy::CircularBuffer | CleanupPolicy::ReadBased => {
            Ok(t_lookback.div_ceil(t_repeat))
        }
        CleanupPolicy::NoCleanup => {
            Err("NoCleanup policy should not call get_sql_cleanup_param".to_string())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::planner::patterns::build_patterns;

    use promql_utilities::ast_matching::PromQLMatchResult;
    use promql_utilities::query_logics::enums::QueryPatternType;

    fn match_query(query: &str) -> (QueryPatternType, PromQLMatchResult) {
        let ast = promql_parser::parser::parse(query).unwrap();
        let patterns = build_patterns();
        for (pt, pattern) in &patterns {
            let result = pattern.matches(&ast);
            if result.matches {
                return (*pt, result);
            }
        }
        panic!("no pattern matched query: {}", query);
    }

    // --- get_effective_repeat ---

    #[test]
    fn effective_repeat_no_step() {
        assert_eq!(get_effective_repeat(300, 0), 300);
    }

    #[test]
    fn effective_repeat_step_smaller_than_t_repeat() {
        assert_eq!(get_effective_repeat(300, 30), 30);
    }

    #[test]
    fn effective_repeat_step_larger_than_t_repeat() {
        assert_eq!(get_effective_repeat(30, 300), 30);
    }

    // --- get_cleanup_param ---

    #[test]
    fn cleanup_param_circular_buffer_spatial_instant_query() {
        let (pt, mr) = match_query("sum(some_metric)");
        assert_eq!(pt, QueryPatternType::OnlySpatial);
        // t_lookback = t_repeat = 300 (OnlySpatial path)
        // effective_repeat = 300 (step=0)
        // ceil((300 + 0) / 300) = 1
        let result = get_cleanup_param(
            CleanupPolicy::CircularBuffer,
            pt,
            &mr,
            300,
            WindowType::Tumbling,
            0,
            0,
        )
        .unwrap();
        assert_eq!(result, 1);
    }

    #[test]
    fn cleanup_param_circular_buffer_spatial_range_query() {
        let (pt, mr) = match_query("sum(some_metric)");
        // t_lookback = t_repeat = 300, effective_repeat = min(300, 30) = 30
        // ceil((300 + 3600) / 30) = ceil(130) = 130
        let result = get_cleanup_param(
            CleanupPolicy::CircularBuffer,
            pt,
            &mr,
            300,
            WindowType::Tumbling,
            3600,
            30,
        )
        .unwrap();
        assert_eq!(result, 130);
    }

    #[test]
    fn cleanup_param_read_based_spatial_instant_query() {
        let (pt, mr) = match_query("sum(some_metric)");
        // lookback_buckets = ceil(300/300) = 1, num_steps = 1 → result = 1
        let result = get_cleanup_param(
            CleanupPolicy::ReadBased,
            pt,
            &mr,
            300,
            WindowType::Tumbling,
            0,
            0,
        )
        .unwrap();
        assert_eq!(result, 1);
    }

    #[test]
    fn cleanup_param_read_based_spatial_range_query() {
        let (pt, mr) = match_query("sum(some_metric)");
        // lookback_buckets = ceil(300/30) = 10, num_steps = 3600/30 + 1 = 121
        // result = 10 * 121 = 1210
        let result = get_cleanup_param(
            CleanupPolicy::ReadBased,
            pt,
            &mr,
            300,
            WindowType::Tumbling,
            3600,
            30,
        )
        .unwrap();
        assert_eq!(result, 1210);
    }

    #[test]
    fn cleanup_param_circular_buffer_temporal_instant_query() {
        let (pt, mr) = match_query("rate(some_metric[5m])");
        assert_eq!(pt, QueryPatternType::OnlyTemporal);
        // t_lookback = 5m = 300s (from [5m] range vector), range_duration=0, step=0
        // effective_repeat = 60, ceil((300 + 0) / 60) = 5
        let result = get_cleanup_param(
            CleanupPolicy::CircularBuffer,
            pt,
            &mr,
            60,
            WindowType::Tumbling,
            0,
            0,
        )
        .unwrap();
        assert_eq!(result, 5);
    }

    #[test]
    fn cleanup_param_no_cleanup_returns_error() {
        let (pt, mr) = match_query("sum(some_metric)");
        let result = get_cleanup_param(
            CleanupPolicy::NoCleanup,
            pt,
            &mr,
            300,
            WindowType::Tumbling,
            0,
            0,
        );
        assert!(result.is_err());
    }

    #[test]
    fn cleanup_param_mismatched_range_and_step_returns_error() {
        let (pt, mr) = match_query("sum(some_metric)");
        // range_duration > 0 but step == 0 is invalid
        let result = get_cleanup_param(
            CleanupPolicy::CircularBuffer,
            pt,
            &mr,
            300,
            WindowType::Tumbling,
            3600,
            0,
        );
        assert!(result.is_err());
    }
}
